runtime/src/leader_bank_notifier.rs

```diff
@@ -0,0 +1,161 @@
+    condvar: Condvar,
+}
+
+/// Leader status state machine for the validator.
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq)]
+enum Status {
+    /// The leader bank is not currently available. Either not initialized, or PoH-completed bank.
+    #[default]
+    StandBy,
+    /// PoH-initiated bank is available.
+    InProgress,
+}
+
+#[derive(Debug, Default)]
+struct SlotAndBankWithStatus {
+    status: Status,
+    slot: Slot,
+    bank: Weak<Bank>,
+}
+
+impl LeaderBankNotifier {
+    /// Set the status to `InProgress` and notify any waiting threads
+    /// if the status was not already `InProgress`.
+    pub fn set_in_progress(&self, bank: &Arc<Bank>) {
+        let mut state = self.state.lock().unwrap();
+        if matches!(state.status, Status::InProgress) {
+            return;
+        }
+
+        *state = SlotAndBankWithStatus {
+            status: Status::InProgress,
+            slot: bank.slot(),
+            bank: Arc::downgrade(bank),
+        };
+        drop(state);
+
+        self.condvar.notify_all();
+    }
+
+    /// Set the status to `StandBy` and notify any waiting threads if
+    ///     1. the status was not already `StandBy` and
+    ///     2. the slot is higher than the current slot (sanity check).
+    pub fn set_completed(&self, slot: Slot) {
+        let mut state = self.state.lock().unwrap();
+        if matches!(state.status, Status::StandBy) || slot < state.slot {
+            return;
+        }
+
+        state.status = Status::StandBy;
+        drop(state);
+
+        self.condvar.notify_all();
+    }
+
+    /// If the status is `InProgress`, immediately return a weak reference to the bank.
+    /// Otherwise, wait up to the `timeout` for the status to become `InProgress`.
+    /// Returns `None` if the timeout is reached.
+    pub fn wait_for_in_progress(&self, timeout: Duration) -> Option<Weak<Bank>> {
+        let state = self.state.lock().unwrap();
+        let (state, wait_timeout_result) = self
+            .condvar
+            .wait_timeout_while(state, timeout, |state| {
+                matches!(state.status, Status::StandBy)
+            })
+            .unwrap();
+
+        (!wait_timeout_result.timed_out()).then(|| state.bank.clone())
+    }
+
+    /// Wait for next notification for a completed leader slot.
+    /// Returns `None` if the timeout is reached
+    pub fn wait_for_next_completed(&self, mut remaining_timeout: Duration) -> Option<Slot> {
+        loop {
+            let start = Instant::now();
+            let state = self.state.lock().unwrap();
+            let (state, result) = self.condvar.wait_timeout(state, remaining_timeout).unwrap();
+            if result.timed_out() {
+                return None;
+            }
+
+            if matches!(state.status, Status::StandBy) {
+                return Some(state.slot);
+            }
+
+            remaining_timeout = remaining_timeout.saturating_sub(start.elapsed());
+        }
```

Comment:
`state` (mutex guard) is immediately dropped here. so, aba problem could be possible.

i mean, indefinite number of cycles of state transition (InProgress => Completed => InProgress => ..) is theoretically possible while this thread has yet to re-enter `.wait_timeout` at the front of next `loop` iteration.

So, I'd suggest putting `let state = ...` outside this loop.

Reply:
I think I may be missing something. We shadow the original `state` by the `MutexGuard` returned from `wait_timeout`. That should be held until the end of the scope (loop iteration), i.e. through this remaining timeout calculation.

Reply:
This got somewhat changed due to https://github.com/solana-labs/solana/pull/30395#discussion_r1144267697, but I am still curious if my previous understanding was wrong